numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -1884,8 +1884,36 @@ gentype_reduce(PyObject *self, PyObject *NPY_UNUSED(args))
     PyTuple_SET_ITEM(ret, 0, obj);
     obj = PyObject_GetAttrString((PyObject *)self, "dtype");
     if (PyArray_IsScalar(self, Object)) {
-        mod = ((PyObjectScalarObject *)self)->obval;
-        PyTuple_SET_ITEM(ret, 1, Py_BuildValue("NO", obj, mod));
+        PyObject *val = ((PyObjectScalarObject *)self)->obval;
+        PyObject *tup = Py_BuildValue("NO", obj, val);
+        if (tup == NULL) {
+            return NULL;
+        }
+        PyTuple_SET_ITEM(ret, 1, tup);
+    }
+    else if (obj && PyDataType_FLAGCHK((PyArray_Descr *)obj, NPY_LIST_PICKLE)) {
+        /* a structured dtype with an object in a field */
+        PyObject *arr = PyArray_FromScalar(self, NULL);
+        if (arr == NULL) {
+            return NULL;
+        }
+        PyArrayIterObject *it = (PyArrayIterObject *)PyArray_IterNew(arr);
+        if (it == NULL) {
+            Py_DECREF(arr);
+            return NULL;
+        }
+        /* arr.item() */
+        PyObject *val = PyArray_GETITEM((PyArrayObject *)arr, it->dataptr);
+        Py_DECREF(it);
+        Py_DECREF(arr);
+        if (val == NULL) {
+            return NULL;
+        }
+        PyObject *tup = Py_BuildValue("NO", obj, val);
```

Comment:
Sorry, you were right to be concerned here - `PyArray_GETITEM` returns a new reference, so as written this leaks `val`
```suggestion
        PyObject *tup = Py_BuildValue("NN", obj, val);
```

Reply:
Or if these strings are annoying, just
```suggestion
PyObject *tup = PyTuple_Pack(2, obj, val);
Py_DECREF(obj);
Py_DECREF(val):
```